
# Computational Genomics Project(02-710)-- Final Project 
---

## Evaluation of  Minimum Distortion Embeddings for Single-Cell RNA sequencing
---
## Introduction 

Organs are comprised of cells, each originating from different lineages and facilitating unique functions. Even within distinct lineages, accumulated evidence shows that cells of the same type can exhibit a striking degree of heterogeneity, such as in the case of  macrophages. Similarly, cancer cells are known to exhibit cellular plasticity, meaning that cancer cells can demonstrate a variety of cell behavior such as stemness, epithelial and mesenchymal transition. Therefore, this creates difficulties in trying to understand biological mechanisms and strategies for therapeutics action. 
Because of the advancement of flow cytometry and fluorescent activated cell sorting (FACS), we are able to distinguish several subtypes in a cell population based on cell-specific markers and fluorescent-labeled using antibodies. This allows us to isolate the cells which express several markers. We can also perform multi-omics analysis to have deeper information of targeted cells after FACS. However, assessing the full spectrum of the transcriptional profile of cells is not possible with these technologies.

The advent of single-cell transcriptomics allows for the molecular profiling at single-cell resolution thereby making it possible to investigate biological mechanisms.  Single-cell RNA sequencing is commonly used  to evaluate mRNA expression profiles of thousands of genes from thousands of individual cells in a high-throughput fashion. In order to analyze such high-dimensional data, dimensionality reduction is performed to enable assessment of different cell clusters present in the data. The latent space induced from such methods can then be used to inform model on downstream tasks, such cell type classification. However, there are many dimensionality reduction methods to choose from, and identifying the one most appropriate for your data can be challenging.

One such method, Minimum Distortion Embeddings (MDE),has recently been developed that claims to be the generalization of traditional dimensionality reduction techniques such as PCA, tSNE, and UMAP. MDE's unique formulation permits its users to specify inductive biases over the distance between samples in the induced latent space. Here, we aim to determine if MDE's flexibility does in fact provide advantages over canonical methods in the downstream task of cell-type classification. We benchmark MDE against PCA, tSNE, and UMAP on a single-cell RNA-seq dataset with 8 known cell types. The embeddings from each dimensionality reduction technique are used as input for a KNN and GMM classifiers, where they are evaluated on a held out test set.

<img width="612" alt="Screen Shot 2021-05-20 at 6 25 41 PM" src="https://user-images.githubusercontent.com/77410526/119056548-dda9ea80-b998-11eb-8df0-e6be1f85399f.png">

## Dataset
For this benchmark, we chose the 10x Genomics Single Cell RNA-seq dataset of Peripheral Blood Mononuclear Cells from a healthy donor generated by Cell Ranger 1.1.0 and Illumina NextSeq 500. This dataset consists of approximately 2,700 single cells with an average of 68,881 reads per cell. The dataset is heterogeneous, with 8 cell types reported.

## Methdology

The flexibility of MDE, afforded by the specification of a distortion function, allows us to evaluate the embeddings produced by each of the 3 main distortion function classes on a cell type clustering task. The evaluation will consist of benchmarking MDE against PCA, tSNE, and UMAP by assessing the accuracy of several clustering algorithms trained on the respective embeddings produced by each method.

## Dimensionality Reduction
We began by splitting out data into a training and test set. Using the training set, we originally computed the embeddings from PCA, tSNE, UMAP, and MDE. However, after training the KNN and GMM classifiers on each of the embeddings, we found that the performance for tSNE, UMAP, and MDE was not at all competitive against PCA. So we proceeded by first computing PCA and then used the first two principal components as input for tSNE, UMAP, MDE. For all dimensionality reduction methods, we used their default parameters.

In computing the 2D latent space for MDE, we used a graph-based distortion function to preserve the k-nearest neighbors (default value of k = 15 was used) of each original data vector. In this way, MDE will preferentially preserve local structure over global structure. Additionally, we specified that the fraction of repulsive and attractive weights over edges in the KNN graph to be set to 1. This means that the number of repulsive and attractive edges is equal. Lastly, we incorporated a constraint to standardized the embeddings as it prevents the vectors from spreading out too much, and enforces the feature columns of the embedding to be uncorrelated.

## Benchmark
Upon computing the principal components for PCA, we fitted tSNE, UMAP, and MDE to the first two principal components of the training dataset. KNN and GMM models were then fit to the embeddings of each respective dimensionality reduction methods. In other words, each dimensionality reduction method had a KNN and GMM classifier fitted to its outputs. We then evaluated the performance of each classifier on a held-out test set. Lastly, the test set predictions for each model were compared to the ground truth class labels to assess performance. 

## Results 

As previously mentioned, when fit to the original data, tSNE, UMAP and MDE performance on the test set was exceptionally poor for both KNN and GMM 

Moving forward, our results for tSNE, UMAP, and MDE fitted to the PCA projected training set show that PCA remains the most competitive for GMM classifiers in terms of test set accuracy. For the KNN classifiers, UMAP had the best performance while PCA and MDE had the second and third best performances, respectively
Inspecting the confusion matrices for our test set predictions illustrates that KNN accuracy for PCA, UMAP, and MDE are comparable. The same cannot be said for the GMM classifiers as each of GMM classifiers struggles in assigning class membership with respect to particular cell types.

The poor performance of GMM on this dataset can perhaps be ascribed to the insufficient cluster separation present within the latent space of the respective dimensionality reduction methods. When cluster separation is poor, the Gaussian densities of the GMM will have overlapping support making cluster assignment challenging. 

Minimum Distortion Embedding, which had been introduced as a competitive alternative to the more popular dimensionality reduction methods like PCA, tSNE, and UMAP, failed to outperform these methods in ScRNA-seq analysis. We observed that the dimensionality reduction methods could separate the major cell types. However, it is difficult to separate the subtypes of the cells such as CD4 and CD8 T cells. This indicates that they  intrinsically share similar gene expression profiles and therefore, it is difficult to distinguish between them. 
In the future, we might apply our modeling approach to more datasets to verify the veracity of our results and also use different variants of MDE like deviation-based MDE. 

## Requirements: 
* Python 3.7+
	- numpy
	- pandas
	- scanpy
	- matplotlib
	- scikit-learn


## collaborators 
Daniel Penaherrera

Keng-Jung Lee
